File: frontend/src/utils/validators.js
// Utilitários de formatação e validação

/**
 * Formata CPF: 000.000.000-00
 */
export const formatCPF = (value) => {
  if (!value) return '';
  
  // Remove tudo que não é número
  const numbers = value.replace(/\D/g, '');
  
  // Limita a 11 dígitos
  const truncated = numbers.slice(0, 11);
  
  // Aplica a máscara
  return truncated
    .replace(/(\d{3})(\d)/, '$1.$2')
    .replace(/(\d{3})(\d)/, '$1.$2')
    .replace(/(\d{3})(\d{1,2})/, '$1-$2');
};

/**
 * Formata Telefone: 00 00000-0000 ou 00 0000-0000
 */
export const formatTelefone = (value) => {
  if (!value) return '';
  
  // Remove tudo que não é número
  const numbers = value.replace(/\D/g, '');
  
  // Limita a 11 dígitos
  const truncated = numbers.slice(0, 11);
  
  // Aplica a máscara
  if (truncated.length <= 10) {
    // Formato: (00) 0000-0000
    return truncated
      .replace(/(\d{2})(\d)/, '$1 $2')
      .replace(/(\d{4})(\d)/, '$1-$2');
  } else {
    // Formato: (00) 00000-0000
    return truncated
      .replace(/(\d{2})(\d)/, '$1 $2')
      .replace(/(\d{5})(\d)/, '$1-$2');
  }
};

/**
 * Remove formatação de CPF e Telefone
 */
export const removeFormat = (value) => {
  if (!value) return '';
  return value.replace(/\D/g, '');
};

/**
 * Formata data do backend (YYYY-MM-DD) para exibição (DD/MM/YYYY)
 * Corrige problema de fuso horário
 */
export const formatDateDisplay = (dateString) => {
  if (!dateString) return '';
  
  // Divide a data em partes (YYYY-MM-DD)
  const [year, month, day] = dateString.split('-');
  
  // Retorna no formato DD/MM/YYYY
  return `${day}/${month}/${year}`;
};

/**
 * Converte data do input (YYYY-MM-DD) para o formato correto
 * sem problemas de fuso horário
 */
export const formatDateForBackend = (dateString) => {
  if (!dateString) return '';
  return dateString; // Já está no formato YYYY-MM-DD
};

/**
 * Valida CPF usando algoritmo de dígitos verificadores
 */
export const validarCPF = (cpf) => {
  // Remove formatação
  const numbers = cpf.replace(/\D/g, '');
  
  // Verifica se tem 11 dígitos
  if (numbers.length !== 11) {
    return false;
  }
  
  // Verifica se todos os dígitos são iguais
  if (/^(\d)\1+$/.test(numbers)) {
    return false;
  }
  
  // Calcula o primeiro dígito verificador
  let soma = 0;
  for (let i = 0; i < 9; i++) {
    soma += parseInt(numbers.charAt(i)) * (10 - i);
  }
  let resto = soma % 11;
  let digito1 = resto < 2 ? 0 : 11 - resto;
  
  if (parseInt(numbers.charAt(9)) !== digito1) {
    return false;
  }
  
  // Calcula o segundo dígito verificador
  soma = 0;
  for (let i = 0; i < 10; i++) {
    soma += parseInt(numbers.charAt(i)) * (11 - i);
  }
  resto = soma % 11;
  let digito2 = resto < 2 ? 0 : 11 - resto;
  
  if (parseInt(numbers.charAt(10)) !== digito2) {
    return false;
  }
  
  return true;
};

/**
 * Valida Email
 */
export const validarEmail = (email) => {
  const regex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
  return regex.test(email);
};

/**
 * Valida Telefone (10 ou 11 dígitos)
 */
export const validarTelefone = (telefone) => {
  const numbers = telefone.replace(/\D/g, '');
  return numbers.length === 10 || numbers.length === 11;
};

File: main/validators.py
import re
from django.core.exceptions import ValidationError
from django.core.validators import RegexValidator

def validar_cpf(cpf: str):
  """
  Valida CPF usando algoritmo de dígitos verificadores.
  Aceita CPF com ou sem formatação (pontos e hífen).
  """
  # Remove caracteres não numéricos
  cpf = re.sub(r'[^0-9]', '', cpf)
  
  # Verifica se tem 11 dígitos
  if len(cpf) != 11:
    raise ValidationError('CPF deve ter 11 dígitos.')
  
  # Verifica se todos os dígitos são iguais (CPF inválido)
  if cpf == cpf[0] * 11:
    raise ValidationError('CPF inválido.')
  
  # Calcula o primeiro dígito verificador
  soma = sum(int(cpf[i]) * (10 - i) for i in range(9))
  resto = soma % 11
  digito1 = 0 if resto < 2 else 11 - resto
  
  if int(cpf[9]) != digito1:
    raise ValidationError('CPF inválido.')
  
  # Calcula o segundo dígito verificador
  soma = sum(int(cpf[i]) * (11 - i) for i in range(10))
  resto = soma % 11
  digito2 = 0 if resto < 2 else 11 - resto
  
  if int(cpf[10]) != digito2:
    raise ValidationError('CPF inválido.')
  
# Filtro regex para formato de CPF
cpf_regex_validator = RegexValidator(
  regex=r'^\d{3}\.\d{3}\.\d{3}-\d{2}$|^\d{11}$',
  message='CPF deve estar no formato 000.000.000-00 ou conter apenas 11 dígitos.',
)

# Filtro regex para formato de telefone
telefone_regex_validator = RegexValidator(
  regex=r'^\d{2}\s\d{4,5}-\d{4}$|^\d{10,11}$',
  message='Telefone deve estar no formato 00 00000-0000 ou conter apenas 10-11 dígitos.',
)

# Filtro regex para formato de email
email_regex_validator = RegexValidator(
  regex=r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$',
  message='Email inválido.',
)
